src/lib/Microsoft.Health.Events/EventCheckpointing/StorageCheckpointClient.cs

```diff
@@ -29,25 +30,53 @@ public class StorageCheckpointClient : ICheckpointClient
         private BlobContainerClient _storageClient;
         private ITelemetryLogger _log;
 
-        public StorageCheckpointClient(StorageCheckpointOptions options, ITelemetryLogger log)
+        public StorageCheckpointClient(IAzureCredentialService credentialService, StorageCheckpointOptions options, ITelemetryLogger log)
         {
             EnsureArg.IsNotNull(options);
-            EnsureArg.IsNotNullOrWhiteSpace(options.BlobPrefix);
-            EnsureArg.IsNotNullOrWhiteSpace(options.BlobStorageConnectionString);
-            EnsureArg.IsNotNullOrWhiteSpace(options.BlobContainerName);
-            EnsureArg.IsNotNullOrWhiteSpace(options.CheckpointBatchCount);
-
             BlobPrefix = options.BlobPrefix;
 
             _lastCheckpointMaxCount = int.Parse(options.CheckpointBatchCount);
             _checkpoints = new ConcurrentDictionary<string, Checkpoint>();
             _lastCheckpointTracker = new ConcurrentDictionary<string, int>();
-            _storageClient = new BlobContainerClient(options.BlobStorageConnectionString, options.BlobContainerName);
+            _storageClient = CreateStorageClient(credentialService, options, log);
             _log = log;
         }
 
         public string BlobPrefix { get; }
 
+        public static BlobContainerClient CreateStorageClient(IAzureCredentialService credentialService, StorageCheckpointOptions options, ITelemetryLogger log)
+        {
+            EnsureArg.IsNotNull(credentialService);
+            EnsureArg.IsNotNull(options);
+            EnsureArg.IsNotNull(log);
+
+            var containerUri = EnsureArg.IsNotNull(options.BlobStorageContainerUri);
+
+            var blobUri = new BlobUriBuilder(containerUri);
+            var tokenCredential = credentialService.GetCredential().TokenCredential;
+            var connectionString = credentialService.GetCredential().ConnectionString;
+
+            if (tokenCredential != null)
+            {
+                return new BlobContainerClient(containerUri, tokenCredential);
+            }
+            else if (!string.IsNullOrWhiteSpace(connectionString))
+            {
+                return new BlobContainerClient(connectionString, blobUri.BlobContainerName);
+            }
+            else
+            {
+                var ex = new Exception($"Unable to create blob container client for {blobUri}");
+                log.LogError(ex);
+                throw ex;
+            }
```

Comment:
Can this be internalized into the credential provider itself?  I.E. creating a storage client for the provider should be a responsibility outside the StorageCheckpointClient and this class should just ask for the provider.  This refactor is a large improvement though!

Reply:
Thanks for the feedback. Went with a factory approach and now pass a storage client in as an argument.